presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java

```diff
@@ -63,53 +66,79 @@ public OutputBufferMemoryManager(long maxBufferedBytes, Supplier<LocalMemoryCont
         this.maxBufferedBytes = maxBufferedBytes;
         this.systemMemoryContextSupplier = Suppliers.memoize(systemMemoryContextSupplier::get);
         this.notificationExecutor = requireNonNull(notificationExecutor, "notificationExecutor is null");
-
-        bufferBlockedFuture = SettableFuture.create();
-        bufferBlockedFuture.set(null);
     }
 
-    public synchronized void updateMemoryUsage(long bytesAdded)
+    public void updateMemoryUsage(long bytesAdded)
     {
-        Optional<LocalMemoryContext> systemMemoryContext = getSystemMemoryContext();
-
-        // If closed is true, that means the task is completed. In that state,
-        // the output buffers already ignore the newly added pages, and therefore
-        // we can also safely ignore any calls after OutputBufferMemoryManager is closed.
-        // If the systemMemoryContext doesn't exist, the task is probably already
-        // aborted, so we can just return (see the comment in getSystemMemoryContext()).
-        if (closed || !systemMemoryContext.isPresent()) {
+        LocalMemoryContext systemMemoryContext = getSystemMemoryContextOrNull();
+        if (systemMemoryContext == null) {
+            // If the systemMemoryContext doesn't exist, the task is probably already
+            // aborted, so we can just return (see the comment in getSystemMemoryContextOrNull()).
             return;
         }
 
-        long currentBufferedBytes = bufferedBytes.addAndGet(bytesAdded);
+        ListenableFuture<?> waitForMemory = null;
+        SettableFuture<?> notifyUnblocked = null;
+        long currentBufferedBytes;
+        synchronized (this) {
+            // If closed is true, that means the task is completed. In that state,
+            // the output buffers already ignore the newly added pages, and therefore
+            // we can also safely ignore any calls after OutputBufferMemoryManager is closed.
+            if (closed) {
+                return;
+            }
+
+            currentBufferedBytes = bufferedBytes.addAndGet(bytesAdded);
+            ListenableFuture<?> blockedOnMemory = systemMemoryContext.setBytes(currentBufferedBytes);
+            if (!blockedOnMemory.isDone()) {
+                if (this.blockedOnMemory != blockedOnMemory) {
+                    this.blockedOnMemory = blockedOnMemory;
+                    waitForMemory = blockedOnMemory; // only register a callback when blocked and the future is different
+                }
+            }
+            else {
+                this.blockedOnMemory = NOT_BLOCKED;
+                if (currentBufferedBytes <= maxBufferedBytes || !blockOnFull.get()) {
```

Comment:
why not use !isBufferFull()?

Reply:
We already know the value of `currentBufferedBytes` and calling `!isBufferFull()` would be another volatile read from the `AtomicLong` inside of the critical section.